File: LATG/Importer/LoadContent.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE ScopedTypeVariables #-}

module LATG.Importer.LoadContent where

import Data.Bifunctor ( Bifunctor(first) )
import Data.Char(toLower)
import qualified Data.Frontmatter as FM
import qualified Data.Aeson as Aeson
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Data.Yaml as Yaml
import qualified LATG.Importer.FileSchema as FSch
import qualified Text.Toml as Toml
import System.FilePath
    ( dropExtension, (</>), takeDirectory, takeExtension )

data EncodedDocument a = EncodedDocument
  { attachedContent :: Maybe (ContentType, BS.ByteString)
  , documentData :: a
  }
  deriving (Show, Eq)

documentOnly :: a -> EncodedDocument a
documentOnly = EncodedDocument Nothing

instance Functor EncodedDocument where
  fmap f (EncodedDocument content doc) = EncodedDocument content (f doc)

data ContentSourceType
  = EmbeddedContent ContentType BS.ByteString
  | FileRef FilePath
  deriving (Show, Eq)

data ContentType
  = PlaintextType
  | MarkdownType
  | HTMLType
  deriving (Show, Eq)

type Result = Either String

data NonDocument = Invalid String | NonDocument 
  deriving (Show, Eq)
type ReadDocumentResult = Either NonDocument

fromEither :: Result a -> ReadDocumentResult a
fromEither (Left err) = Left $ Invalid err
fromEither (Right x) = Right x

toEither :: ReadDocumentResult a -> Result (Maybe a)
toEither (Left (Invalid err)) = Left err
toEither (Left NonDocument) = Right Nothing
toEither (Right x) = Right $ Just x

invalid :: String -> ReadDocumentResult a
invalid err = Left $ Invalid err

load :: FilePath -> IO (ReadDocumentResult (FSch.GenericDocument, T.Text))
load path = do
  rawDoc <- BL.readFile path

  case readDocument (takeExtension path) rawDoc of 
    Left e -> return $ Left e
    Right doc -> case extractContentSource path $ fmap FSch.content doc of
      Left e -> return $ invalid e
      Right cs -> do
        result <- loadContentSource cs
        case result of 
          Left e -> return $ invalid e
          Right (contentType, contentRaw) -> do
            transformed <- transformContent contentType contentRaw 
            return $ case transformed of
              Left e -> invalid e
              Right transformed' -> Right (documentData doc, transformed')

readDocument :: Aeson.FromJSON a => String -> BL.ByteString -> ReadDocumentResult (EncodedDocument a)
readDocument extension content = 
  case extension of
    ".md" -> markdown
    ".json" -> fromEither json
    ".yml" -> fromEither yaml
    ".yaml" -> fromEither yaml
    ".toml" -> fromEither toml
    _ -> Left NonDocument
  where
    json = documentOnly <$> Aeson.eitherDecode content 

    markdown = case FM.parseFrontmatter $ BL.toStrict content of
      FM.Done body front -> case Yaml.decodeEither' front of
        Left err -> invalid $ show err
        Right x -> Right $ EncodedDocument (Just (MarkdownType, body)) x
      _ -> Left NonDocument

    yaml = documentOnly <$> first show (Yaml.decodeEither' $ BL.toStrict content)

    toml = do
      table <- first show $ Toml.parseTomlDoc "" $ TE.decodeUtf8 $ BL.toStrict content 
      case Aeson.fromJSON $ Aeson.toJSON table of
        Aeson.Error err -> Left $ show err
        Aeson.Success x -> Right $ documentOnly x

extractContentSource :: FilePath -> EncodedDocument (Maybe FSch.Content) -> Result ContentSourceType
extractContentSource _ (EncodedDocument (Just (contentType, contentRaw)) contentData) = case contentData of
  Nothing -> Right $ EmbeddedContent contentType contentRaw
  Just _ -> Left "Embedded content cannot reference other sources"
extractContentSource path (EncodedDocument Nothing contentData) = case contentData of
  Nothing -> withoutExtension
  Just (FSch.EmbeddedPlaintext text) -> Right $ EmbeddedContent PlaintextType $ TE.encodeUtf8 text
  Just (FSch.FileRef srcMaybe _) -> case srcMaybe of
    Nothing -> withoutExtension
    Just src -> Right $ FileRef $ takeDirectory path </> src
  where 
    withoutExtension = Right $ FileRef $ dropExtension path

loadContentSource :: ContentSourceType -> IO (Result (ContentType, BS.ByteString))
loadContentSource (EmbeddedContent contentType body) = pure $ Right (contentType, body)
loadContentSource (FileRef path) = do
  content <- BS.readFile path

  return $ case map toLower $ takeExtension path of
    ".html" -> Right (HTMLType, content)
    ".htm" -> Right (HTMLType, content)
    ".md" -> Right (MarkdownType, content)
    ".txt" -> Right (PlaintextType, content)
    _ -> Left $ "Unsupported content file " ++ path

transformContent :: ContentType -> BS.ByteString -> IO (Result T.Text)
transformContent _ raw = pure $ Right $ TE.decodeUtf8 raw  -- TODO use pandoc

File: latg/importer/src/load_content.rs
use std::{
    ffi::OsStr,
    fs,
    path::{Path, PathBuf},
};

use gray_matter::{engine::YAML, Matter, Pod};
use serde::{de::DeserializeOwned, Deserialize};

use crate::file_schema::{self, Document};

struct Content {
    content_type: ContentType,
    raw: String,
}

struct DocumentWithPossiblyEmbeddedContent<T> {
    embedded_content: Option<Content>,
    document: T,
}

enum ContentSourceType<'a> {
    Embedded(ContentType, &'a str),
    FileRef(PathBuf),
}

enum ContentType {
    Plaintext,
    Markdown,
    Html,
    // Jupyter,
}

enum NonDocument {
    Invalid(String),
    NonDocument,
}

struct TransformedContent {
    html: String,
    assets: Vec<String>, // TODO
}

type ReadDocumentResult<T> = Result<T, NonDocument>;
type ContentResult<T> = Result<T, String>;

fn read_document<T: DeserializeOwned>(
    extension: &OsStr,
    file_content: &str,
) -> ReadDocumentResult<DocumentWithPossiblyEmbeddedContent<T>> {
    match extension.to_string_lossy().as_ref() {
        "md" | "markdown" => {
            let entity = Matter::<YAML>::new().parse(file_content);
            let frontmatter: Option<Pod> = entity.data;

            if let Some(document) = frontmatter {
                Ok(DocumentWithPossiblyEmbeddedContent {
                    document: document.deserialize()?,
                    embedded_content: Some(entity.content),
                })
            } else {
                Err(NonDocument::NonDocument)
            }
        }
        "yml" | "yaml" => Ok(DocumentWithPossiblyEmbeddedContent {
            embedded_content: None,
            document: serde_yaml::from_str(file_content)?,
        }),
        "json" => Ok(DocumentWithPossiblyEmbeddedContent {
            embedded_content: None,
            document: serde_json::from_str(file_content)?,
        }),
        "toml" => Ok(DocumentWithPossiblyEmbeddedContent {
            embedded_content: None,
            document: toml::from_str(file_content)?,
        }),
        _ => Err(NonDocument::NonDocument),
    }
}

fn extract_content_source(
    path: &Path,
    content_info: DocumentWithPossiblyEmbeddedContent<Option<file_schema::Content>>,
) -> ContentResult<ContentSourceType> {
    if let Some(embedded_content) = content_info.embedded_content {
        return if let Some(_) = content_info.document {
            Err("Documents with embedded content cannot reference other sources".to_string())
        } else {
            Ok(ContentSourceType::Embedded(
                embedded_content.content_type,
                &embedded_content.raw,
            ))
        };
    }

    fn ref_without_extension(path: &Path) -> ContentResult<ContentSourceType> {
        if let Some(stem) = path.file_stem() {
            Ok(ContentSourceType::FileRef(PathBuf::from(stem)))
        } else {
            Err(format!(
                "Could not find file name for {}",
                path.to_string_lossy()
            ))
        }
    }

    match content_info.document {
        None => ref_without_extension(path),
        Some(file_schema::Content::EmbeddedPlaintext(plaintext)) => Ok(
            ContentSourceType::Embedded(ContentType::Plaintext, &plaintext),
        ),
        Some(file_schema::Content::FileRef { src, .. }) => {
            if let Some(src) = src {
                let mut path = PathBuf::from(path);
                path.pop();
                path.push(src);
                Ok(ContentSourceType::FileRef(path))
            } else {
                ref_without_extension(path)
            }
        }
    }
}

fn load_content_source(source_type: ContentSourceType) -> ContentResult<Content> {
    Ok(match source_type {
        ContentSourceType::Embedded(content_type, raw) => Content {
            content_type,
            raw: raw.to_string(),
        },
        ContentSourceType::FileRef(path) => match path.extension() {
            Some(ext) => match ext.to_string_lossy().as_ref() {
                "html" | "htm" => Content {
                    content_type: ContentType::Html,
                    raw: fs::read_to_string(path)?,
                },
                "md" | "markdown" => Content {
                    content_type: ContentType::Markdown,
                    raw: fs::read_to_string(path)?,
                },
                "txt" => Content {
                    content_type: ContentType::Plaintext,
                    raw: fs::read_to_string(path)?,
                },
                _ => Err(format!(
                    "Unsupported content at path {}",
                    path.to_string_lossy()
                ))?,
            },
            None => Err(format!("No extension on file {}", path.to_string_lossy()))?,
        },
    })
}

fn transform_content(content: Content) -> ContentResult<TransformedContent> {
    match content.content_type {
        ContentType::Plaintext => Ok(TransformedContent {
            html: format!("<pre>{}</pre>", html_escape::encode_text(&content.raw)),
            assets: vec![],
        }),
        ContentType::Markdown => todo!(),
        ContentType::Html => Ok(TransformedContent {
            html: content.raw,
            assets: vec![],
        }),
    }
}
